File: tests/test_programs/single_step_test.c
#include <windows.h>

// EXCEPTION_SINGLE_STEP / STATUS_SINGLE_STEP
#define SINGLE_STEP_CODE 0x80000004

volatile int g_caught = 0;

// The program deliberately raises a single-step exception itself (as an
// anti-debug check or self-tracer would). Under a debugger this surfaces as a
// generic Exception event with code 0x80000004 (the server's "unexpected SS"
// path), NOT as a debugger-initiated step. Whether the program's own SEH
// handler runs depends on how the debugger continues it.
__declspec(noinline)
void raise_and_catch(void) {
    __try {
        RaiseException(SINGLE_STEP_CODE, 0, 0, NULL);
        // If DBG_CONTINUE swallows the exception, execution continues here
        // and the __except block is skipped.
    } __except(EXCEPTION_EXECUTE_HANDLER) {
        g_caught = 1;
    }
}

int main(void) {
    raise_and_catch();
    // Exit 0 = exception passed to application (SEH handler ran)
    // Exit 1 = exception swallowed by debugger (bug)
    return g_caught == 1 ? 0 : 1;
}

File: tests/single_step_test.rs
#![cfg(windows)]

mod common;

use common::{TestServer, get_test_program_path};
use joybug2::protocol_io::{DebugSession, ExceptionAction};

// EXCEPTION_SINGLE_STEP / STATUS_SINGLE_STEP.
const SINGLE_STEP_CODE: u32 = 0x80000004;

fn test_exe_path() -> String {
    get_test_program_path("single_step_test")
}

/// A single-step exception raised by the debuggee itself must be treatable like
/// any other exception. With HandledByDebugger (DBG_CONTINUE) the exception is
/// swallowed, the program's SEH handler never runs, and it exits with code 1.
#[test]
fn test_single_step_handled_by_debugger() {
    let server = TestServer::spawn();
    let server_addr = server.address().to_string();
    let test_exe = test_exe_path();

    struct TestState {
        exit_code: Option<u32>,
    }

    let final_state = DebugSession::new(
        TestState { exit_code: None },
        Some(server_addr.as_str()),
    )
    .expect("Failed to connect to debug server")
    .set_exception_action(
        SINGLE_STEP_CODE,
        ExceptionAction::HandledByDebugger,
        ExceptionAction::Stop,
    )
    .on_process_exited(|session, _pid, exit_code| {
        session.state.exit_code = Some(exit_code);
        Ok(())
    })
    .launch(test_exe)
    .expect("Debug session failed");

    assert_eq!(
        final_state.exit_code,
        Some(1),
        "Expected exit code 1 (single-step exception swallowed by debugger), got {:?}",
        final_state.exit_code
    );
}

/// The reported case: "Go (Pass Exception)" on a program-raised single-step
/// exception. With PassToApplication (DBG_EXCEPTION_NOT_HANDLED) the exception
/// is delivered to the program, its SEH handler runs, and it exits with code 0.
#[test]
fn test_single_step_passed_to_application() {
    let server = TestServer::spawn();
    let server_addr = server.address().to_string();
    let test_exe = test_exe_path();

    struct TestState {
        exit_code: Option<u32>,
    }

    let final_state = DebugSession::new(
        TestState { exit_code: None },
        Some(server_addr.as_str()),
    )
    .expect("Failed to connect to debug server")
    .set_exception_action(
        SINGLE_STEP_CODE,
        ExceptionAction::PassToApplication,
        ExceptionAction::Stop,
    )
    .on_process_exited(|session, _pid, exit_code| {
        session.state.exit_code = Some(exit_code);
        Ok(())
    })
    .launch(test_exe)
    .expect("Debug session failed");

    assert_eq!(
        final_state.exit_code,
        Some(0),
        "Expected exit code 0 (single-step exception passed to application), got {:?}",
        final_state.exit_code
    );
}
